src/adr/0051-pulsar-mailqueue.md

```diff
@@ -0,0 +1,197 @@
+ referenced by the Pulsar MailQueue, see [JIRA-XXXX](TODO).
+ - The event bus (described in [ADR 37](0037-eventbus.md)) would benefit from a Pulsar implementation, replacing the 
+ existing RabbitMQ one (described in [ADR-38](0038-distributed-eventbus.md)). See [JIRA-XXXX](TODO).
+ - While being less critical, a task manager implementation would be needed as well to replace the RabbitMQ one
+ described in [ADR 2](0002-make-taskmanager-distributed.md) [ADR 3](0003-distributed-workqueue.md) 
+ [ADR 4](0004-distributed-tasks-listing.md) [ADR 5](0005-distributed-task-termination-ackowledgement.md) 
+ [ADR 6](0006-task-serialization.md) [ADR 7](0007-distributed-task-cancellation.md) 
+ [ADR 8](0008-distributed-task-await.md), eventually allowing to drop the RabbitMQ technology all-together.
+
+We could then create a new artifact relying solely on Pulsar, and deprecate the RabbitMQ based artifact.
+
+Priorities are not yet supported by the current implementation. See [JIRA-XXXX](TODO).
+
+A bug regarding clear not purging delayed messages had been 
+[reported](https://github.com/apache/james-project/pull/808#discussion_r780162174) as well.
+
+A broader adoption of Pulsar would benefit from performance insights.
+
+This work could be continued, for instance under the form of a Google Summer of Code for 2022.
+
+## Technical details
+
+[[This section requires a deep review]]
+
+[Akka](https://akka.io/) actor system is used in single node mode as a processing framework.
+
+The MailQueue relies on the following topology:
+
+ - out topic : emails are first enqueued here.
+ - scheduled topic: emails that are delayed are first enqueued there.
+ - filter topic: Deletions (name, sender, recipients) prior a given sequence are synchronized between nodes using this topic.
+ 
+[[Question]]
+```
+https://github.com/apache/james-project/pull/808#discussion_r780181972
+
+Likely a stupid question but why using a separate topic for delayed messages?
+
+My take is that it limits the out-of-order impact and yields better performance.
+
+Do you confirm?
+```
+
+Upon enqueue, the blobs are first saved, then the Pulsar message payload is generated and published to the relevant 
+topic (out or scheduled).
+
+Scheduled messages have their `deliveredAt` property set to the desired value. When the delay is 
+expired, the message will be consumed and thus moved to the out topic. Flushes simply copy content of the scheduled
+topic to the out topic then reset the offset of the scheduled queue, atomically. Expired filters are removed.
+
+[[Question]]
+```
+https://github.com/apache/james-project/pull/808#discussion_r780170243
+
+> Expired filters are removed.
+
+Given we could get emails with arbitrary delays, invalidating filters likely lead to bugs, as the underlying emails can 
+be out of order.
+
+Maybe it is safer to get rid of this optimisation. Correct?
+```
```

Comment:
@mbaechler @jeantil  opinion on this?

Reply:
https://github.com/apache/james-project/pull/831

It's worse. Deletes fails on delayed items... Where they are likely the most usefull...

Reply:
I don't think keeping expired filters in memory is useful.
If we don't expire the filters at all they will continue to apply to all future email which is not how I understood the feature. As [I wrote](https://github.com/apache/james-project/pull/808#discussion_r780180667) in response to the same comment on the PR: 
> What we could look into is purging all scheduled messages that have been scheduled before the call to remove(), it requires a bit of rework but should be doable.